arangod/RestHandler/RestAdminClusterHandler.cpp

```diff
@@ -2135,6 +2143,413 @@ RestStatus RestAdminClusterHandler::handleRebalanceShards() {
           }));
 }
 
+namespace {
+struct RebalanceOptions {
+  std::uint64_t version;
+  std::size_t maximumNumberOfMoves;
+  bool leaderChanges;
+  bool moveLeaders;
+  bool moveFollowers;
+  double piFactor;
+  std::vector<DatabaseID> databasesExcluded;
+};
+
+template<class Inspector>
+auto inspect(Inspector& f, RebalanceOptions& x) {
+  return f.object(x).fields(
+      f.field("version", x.version),
+      f.field("maximumNumberOfMoves", x.maximumNumberOfMoves)
+          .fallback(std::size_t{1000}),
+      f.field("leaderChanges", x.leaderChanges).fallback(true),
+      f.field("moveLeaders", x.moveLeaders).fallback(false),
+      f.field("moveFollowers", x.moveFollowers).fallback(false),
+      f.field("piFactor", x.piFactor).fallback(256e6),
+      f.field("databasesExcluded", x.databasesExcluded)
+          .fallback(std::vector<DatabaseID>{}));
+};
+
+}  // namespace
+
+RestAdminClusterHandler::MoveShardCount
+RestAdminClusterHandler::countAllMoveShardJobs() {
+  auto& cache = server().getFeature<ClusterFeature>().agencyCache();
+
+  auto const countMoveShardsInSlice = [](VPackSlice slice) -> std::size_t {
+    std::size_t count = 0;
+    for (auto const& [key, job] : VPackObjectIterator(slice)) {
+      if (job.get("type").isEqualString("moveShard")) {
+        count += 1;
+      }
+    }
+
+    return count;
+  };
+
+  auto const countMoveShardsAt = [&](std::string const& path) -> std::size_t {
+    auto [query, index] = cache.get(path);
+    return countMoveShardsInSlice(query->slice());
+  };
+  return MoveShardCount{
+      .todo = countMoveShardsAt("Target/ToDo"),
+      .pending = countMoveShardsAt("Target/Pending"),
+  };
+}
+
+RestStatus RestAdminClusterHandler::handleRebalanceGet() {
+  if (request()->suffixes().size() != 1) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                  "expect GET /_admin/cluster/rebalance");
+    return RestStatus::DONE;
+  }
+
+  auto [todo, pending] = countAllMoveShardJobs();
+
+  VPackBuilder builder;
+  auto p = collectRebalanceInformation({});
+  auto leader = p.computeLeaderImbalance();
+  auto shard = p.computeShardImbalance();
+  {
+    VPackObjectBuilder ob(&builder);
+    builder.add(VPackValue("leader"));
+    velocypack::serialize(builder, leader);
+    builder.add(VPackValue("shards"));
+    velocypack::serialize(builder, shard);
+    builder.add("pendingMoveShards", VPackValue(pending));
+    builder.add("todoMoveShards", VPackValue(todo));
+  }
+
+  generateOk(rest::ResponseCode::OK, builder.slice());
+  return RestStatus::DONE;
+}
+
+RestStatus RestAdminClusterHandler::handleRebalanceExecute() {
+  if (request()->requestType() != rest::RequestType::POST) {
+    generateError(rest::ResponseCode::METHOD_NOT_ALLOWED,
+                  TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+    return RestStatus::DONE;
+  }
+
+  auto batch = velocypack::deserialize<std::vector<MoveShardDescription>>(
```

Comment:
Lets make the API as follows: One sends a document with `moves` attribute being a list of moves, and set the `version` attribute to 1. Then we can extend the API a bit better. If we just get a list of moves, we have no wiggle room for further options.

Reply:
ok.